presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownSpatialProjections.java

```diff
@@ -0,0 +1,201 @@
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.Assignments;
+import com.facebook.presto.sql.planner.plan.FilterNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.NodeRef;
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.google.common.base.Verify;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.matching.Capture.newCapture;
+import static com.facebook.presto.sql.analyzer.ExpressionAnalyzer.getExpressionTypes;
+import static com.facebook.presto.sql.planner.ExpressionNodeInliner.replaceExpression;
+import static com.facebook.presto.sql.planner.SymbolsExtractor.extractUnique;
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.REPLICATED;
+import static com.facebook.presto.sql.planner.plan.Patterns.filter;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+import static com.facebook.presto.util.SpatialJoinUtils.extractSupportedSpatialFunctions;
+import static java.util.Collections.emptyList;
+
+/**
+ * Applies to broadcast spatial joins using ST_Contains and ST_Intersects functions.
+ * <p>
+ * Applies only if all of the following conditions are met:
+ * - each argument of the spatial function is an expression using at least one of the symbols
+ *      from the join relations
+ * - each argument's expression uses symbols from only one side of the join
+ * - different arguments use symbols from different sides of the join
+ * <p>
+ * Replaces cross join node with a filter on top with a single spatial join node.
+ * <p>
+ * Pushes non-trivial expressions of the spatial function arguments into projections on top of
+ * join child nodes.
+ * <p>
+ * For example, rewrites ST_Contains(ST_GeometryFromText(a.wkt), ST_Point(b.longitude, b.latitude)) join
+ * as ST_Contains(st_geometryfromtext, st_point) with st_geometryfromtext -> 'ST_GeometryFromText(a.wkt)' and
+ * st_point -> 'ST_Point(b.longitude, b.latitude)' projections on top of child nodes.
+ */
+public class PushDownSpatialProjections
+        implements Rule<FilterNode>
+{
+    private static final Capture<JoinNode> JOIN = newCapture();
+    private static final Pattern<FilterNode> PATTERN = filter()
+            .with(source().matching(join().capturedAs(JOIN).matching(node -> node.isCrossJoin() && node.getDistributionType().map(REPLICATED::equals).orElse(false))));
```

Comment:
should not be the other way round that spatial join enforce join to be replicated? Now if join is replicated then it can become spatial.

Reply:
>> Now if join is replicated then it can become spatial.

This is the behavior I want. Replicated join is only possible if build side is sufficiently small. Currently, we rely on the user to determine whether this is the case or not.